ww4ff/run/train.py

```diff
@@ -10,93 +9,105 @@
+    ww_dev_neg_ds = ww_dev_ds.filter(lambda x: not x.compute_frame_labels(args.vocab), clone=True)
+    ww_test_pos_ds = ww_test_ds.filter(lambda x: x.compute_frame_labels(args.vocab), clone=True)
+    ww_test_neg_ds = ww_test_ds.filter(lambda x: not x.compute_frame_labels(args.vocab), clone=True)
 
-    workspace = Workspace(Path(args.workspace), delete_existing=not args.eval)
-    workspace.write_args(args)
-    workspace.write_setting(SETTINGS)
-    writer = workspace.summary_writer
+    ww_dev_pos_ds = WakeWordEvaluationDataset(ww_dev_pos_ds, wind_sz, stri_sz, num_labels - 1)
+    ww_dev_neg_ds = WakeWordEvaluationDataset(ww_dev_neg_ds, wind_sz, stri_sz, num_labels - 1)
+    ww_test_pos_ds = WakeWordEvaluationDataset(ww_test_pos_ds, wind_sz, stri_sz, num_labels - 1)
+    ww_test_neg_ds = WakeWordEvaluationDataset(ww_test_neg_ds, wind_sz, stri_sz, num_labels - 1)
 
     device = torch.device(SETTINGS.training.device)
     std_transform = StandardAudioTransform(sr).to(device).eval()
     zmuv_transform = ZmuvTransform().to(device)
-    train_comp = compose(trim,
-                         partial(random_slice, max_window_size=int(SETTINGS.training.max_window_size_seconds * sr)),
-                         TimeshiftTransform().train(),
-                         TimestretchTransform().train(),
-                         NoiseTransform().train(),
-                         batchify)
-    prep_dl = StandardAudioDataLoaderBuilder(ww_train_ds, collate_fn=compose(trim, batchify)).build(1)
+    batchifier = WakeWordBatchifier(num_labels - 1,
+                                    window_size_ms=int(SETTINGS.training.max_window_size_seconds * 1000))
+    train_comp = compose(NoiseTransform().train(), batchifier)
+    prep_dl = StandardAudioDataLoaderBuilder(ww_train_ds, collate_fn=batchify).build(1)
+    prep_dl.shuffle = True
     train_dl = StandardAudioDataLoaderBuilder(ww_train_ds, collate_fn=train_comp).build(SETTINGS.training.batch_size)
 
     model = find_model(args.model)().to(device)
     params = list(filter(lambda x: x.requires_grad, model.parameters()))
-    writer.add_scalar('Meta/Parameters', sum(p.numel() for p in params))
-    optimizer = AdamW(params, SETTINGS.training.learning_rate)
+    optimizer = AdamW(params, SETTINGS.training.learning_rate, weight_decay=SETTINGS.training.weight_decay)
     logging.info(f'{sum(p.numel() for p in params)} parameters')
     criterion = nn.CrossEntropyLoss()
 
-    for batch in tqdm(prep_dl, desc='Constructing ZMUV'):
-        batch.to(device)
-        zmuv_transform.update(std_transform(batch.audio_data))
-
+    if (ws.path / 'zmuv.pt.bin').exists():
+        zmuv_transform.load_state_dict(torch.load(str(ws.path / 'zmuv.pt.bin')))
+    else:
+        for idx, batch in enumerate(tqdm(prep_dl, desc='Constructing ZMUV')):
+            batch.to(device)
+            zmuv_transform.update(std_transform(batch.audio_data))
+            if idx == 2000:  # TODO: quick debugging, remove later
```

Comment:
should we keep it still?

Reply:
Yeah, just for speed. Also, CLT says 2000 should be fine.